Engine/DataFeeds/Enumerators/CorporateEventBaseEnumerator.cs

```diff
@@ -23,21 +23,12 @@
 namespace QuantConnect.Lean.Engine.DataFeeds.Enumerators
 {
     /// <summary>
-    /// Base class for corporate event enumerators. Will call the <see cref="CheckNewEvent"/>
-    /// implementation each time there is a new tradable day. Will merge new events produced with
-    /// underlying enumerator data.
+    /// Base class for corporate event enumerators. Will call the <see cref="GetCorporateEvents"/>
+    /// implementation each time there is a new tradable day.
     /// </summary>
     public abstract class CorporateEventBaseEnumerator : IEnumerator<BaseData>
     {
-        private readonly IEnumerator<BaseData> _enumerator;
         private readonly Queue<BaseData> _auxiliaryData;
```

Comment:
It seems a little suspect that this enumerator requires a queue. The general pattern w/ the (newer) enumerator stacks is that they're very lazy and have very minimal memory footprint . I was thinking this would hold open a file stream and advance the stream one line at a time. Also, if we implement `<CorporateEventType>.Reader` and `<CorporateEventType>.GetSource` then we should be able to use the default infrastructure for reading `BaseData` instances (such as the, IIRC, [`TextSubscriptionDataSourceReader`](https://github.com/QuantConnect/Lean/blob/master/Engine/DataFeeds/TextSubscriptionDataSourceReader.cs)

Reply:
The queue is required because the event is not produced aligned with the `MoveNext()` call of the enumerator. Why is it not aligned?
- It depends on the `ITradableDatesNotifier` to notify a new date event. This could be replaced if each corporate event type had its own tradable date enumerator but that does not seem like a good idea (today we would be looping 5x times on the tradable dates separately). Also the tradable date is advanced based on the price data.
- It depends on the `ITradableDateNotifier` `Price` value, why? The price value is still deep in the stack (`SubscriptionDataReader`) and has not reached the `Synchronizer`/`algoManager`, so it can't be accessed by the corporate event type.

We also need the queue because the corporate event producer might generate more than one event per tradable date, the `IEnumerable<BaseData>` return type

We might be able to have a queue holding the `NewTradableDateEventArgs` and process them on demand (still new a queue to hold the events): This one _might_ open a window for bugs, since the system status will be different and the corporate event producer might not realize/be aware of it.

We could get rid of the `ITradableDatesNotifier` if the `SubscriptionDataReader` hold the corporate event enumerators an drove them (`MoveNext()`) when required, but this would be like the current implementation and still require a queue at the `SubscriptionDataReader` level